### PR TITLE
Add jdbc metadata cache flush procedure

### PR DESCRIPTION
**Pull request by aczajkowski**
Extracted from https://github.com/trinodb/trino/pull/10251 for clearance.

### CODE REVIEW

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java [hunk 1]
```diff
@@ -34,9 +40,29 @@ protected QueryRunner createQueryRunner()
                 .put("metadata.cache-missing", "true")
                 .put("allow-drop-table", "true")
                 .build();
+        this.h2SqlExecutor = new JdbcSqlExecutor(properties.get("connection-url"), new Properties());
         return createH2QueryRunner(REQUIRED_TPCH_TABLES, properties);
     }
 
+    @Test
+    public void testFlushJdbcMetadataCacheProcedureCallable()
```
Comment (hashhar) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:48]: Add another test to test the remote identifier cache flush.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java [hunk 2]
```diff
@@ -34,9 +40,29 @@ protected QueryRunner createQueryRunner()
...
+    {
+        assertUpdate("CREATE TABLE cached AS SELECT * FROM nation", 25);
+        assertQuerySucceeds("SELECT name, regionkey FROM cached");
+
+        //Rename column outside Trino
```
Comment (hashhar) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:53]: nit: `// Rename` (here and other comments)

**Review by hashhar — COMMENTED** [review]
LGTM % comment

I confirmed all caches (in base-jdbc) are covered by this.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java (hunk 1 above)
Comment (aczajkowski) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:48]: I was trying to add it. But identifier mapping is only caching schema and table names.
Also cache key is based on name it self which causes it to resolve when new un-existing value is asked.
So we just resolve it again. 
I tried it to check on errors like. Caching table name, renaming it outside of Trino, and making query to it (we should get error from target DB), and then asking again for un-existing table so we should get error from Trino as cache would be all-ready update. But errors are the same :( so I can't recognise if cache was flushed or not. 
Any suggestions, how we could approach it ?

Comment (hashhar) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:48]: > Also cache is implement to reload when value is not found.

See `metadata.cache-missing`. Enabling it caches cache misses too.

Comment (aczajkowski) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java]: Thx will try it out.

Comment (aczajkowski) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java]: @hashhar `metadata.cache-missing` is implemented only for metadata cache in JdbcClient.
Then one implemented in CachingIdentifierMapping is not using such behaviour.
Also maybe I'm missing something but i think this step:
> Then you can create the missing table using H2 (use mixed case to ensure remote name cache gets used).

will not fill remote names cache (CachingIdentifierMapping) as we will make it no H2 omitting Trino and it's caches.

Comment (hashhar) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java]: The test isn't working because TestingH2JdbcModule doesn't install the IdentifierMappingModule and is using the no-cache DefaultIdentifierMapping.

I also notice now that the module also doesn't bind a CachingJdbcClient. Only a plain TestingH2JdbcClient is created.

Comment (hashhar) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java]: You can verify this by trying to query a lowercase table in H2.

e.g. In H2 `CREATE TABLE "lowercase"` and then you won't be able to query it from Trino.

I think in the end a test like below should pass.

```java
    @Test
    public void testFlushIdentifierCacheProcedureFlushCache()
    {
        assertUpdate("CREATE TABLE cached_name AS SELECT * FROM nation", 25);
        String query = "SELECT name, regionkey FROM cached_name";
        assertQuerySucceeds(query);

        // Rename to lowercase name outside Trino (H2 stores unquoted names as uppercase)
        h2SqlExecutor.execute("ALTER TABLE tpch.cached_name RENAME TO tpch.\"cached_name\"");

        // Should fail as Trino has old identifier cached
        assertThatThrownBy(() -> getQueryRunner().execute(query))
                .hasMessageMatching("(?s)Table \"CACHED_NAME\" not found.*");

        // Should succeed after flushing Trino cache
        getQueryRunner().execute(getSession(), "CALL system.flush_metadata_cache()");
        assertQuerySucceeds(query);
    }
```

Comment (aczajkowski) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java]: From beginning I'm trying to implement test in https://github.com/trinodb/trino/blob/master/plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseCaseInsensitiveMappingTest.java as there are test covering test scenarios with IdentifierMapping . 
Let me try doing as you suggest:
- wired IdentifierMappingModule in TestJdbcCachingConnectorSmokeTest
- add test as above

Comment (aczajkowski) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java]: @hashhar thank you for your help - much appreciated.
This is the trick i was missing `Rename to lowercase name outside Trino (H2 stores unquoted names as uppercase)`  
To keeps tests together I've modified TestingH2JdbcClient to use IdentifierMapping from Guice context instead of always default and added test together in TestJdbcCachingConnectorSmokeTest

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/procedure/FlushJdbcMetadataCacheProcedure.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                "system",
+                "flush_metadata_cache",
+                ImmutableList.of(),
+                FLUSH_JDBC_METADATA_CACHE.bindTo(this));
+    }
+
+    public void flushMetadataCache()
+    {
+        cachingJdbcClient.flushCache();
+        cachingIdentifierMapping.ifPresent(cache -> cache.flushCache());
```
Comment (hashhar) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/procedure/FlushJdbcMetadataCacheProcedure.java:62]: nit:
```suggestion
        cachingIdentifierMapping.ifPresent(CachingIdentifierMapping::flushCache);
```

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java [hunk 4]
```diff
@@ -16,27 +16,79 @@
...
             throws Exception
     {
+        Path mappingFile = createRuleBasedIdentifierMappingFile();
         Map<String, String> properties = ImmutableMap.<String, String>builder()
                 .putAll(TestingH2JdbcModule.createProperties())
                 .put("metadata.cache-ttl", "10m")
                 .put("metadata.cache-missing", "true")
+                .put("case-insensitive-name-matching", "true")
+                .put("case-insensitive-name-matching.config-file", mappingFile.toFile().getAbsolutePath())
+                .put("case-insensitive-name-matching.config-file.refresh-period", "100s")
```
Comment (hashhar) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:46]: Remove these 2 since you aren't actually using any "rules" within this test class.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java [hunk 5]
```diff
@@ -16,27 +16,79 @@
...
+
+        // Rename column outside Trino
+        h2SqlExecutor.execute("ALTER TABLE tpch.cached ALTER COLUMN regionkey RENAME TO renamed");
+
+        String renamedColumnQuery = "SELECT name, renamed FROM cached";
+        // Should fail as Trino has old metadata cached
+        assertThatThrownBy(() -> getQueryRunner().execute(renamedColumnQuery))
+                .hasMessageMatching(".*Column 'renamed' cannot be resolved");
+
+        // Should success after flushing Trino JDBC metadata cache
```
Comment (hashhar) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:67]:
```suggestion
        // Should succeed after flushing Trino JDBC metadata cache
```

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/IdentifierMappingModule.java [hunk 6]
```diff
@@ -57,6 +58,7 @@ protected void setup(Binder binder)
         binder.bind(DefaultIdentifierMapping.class).in(SINGLETON);
 
         MappingConfig config = buildConfigObject(MappingConfig.class);
+        newOptionalBinder(binder, CachingIdentifierMapping.class);
```
Comment (hashhar) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/IdentifierMappingModule.java:61]: Do the changes in this class deserve a comment. It's not obvious to me why this change was needed (at least until now).

File: docs/src/main/sphinx/connector/jdbc-procedures.fragment [hunk 7]
```diff
@@ -0,0 +1,6 @@
+Procedures
+^^^^^^^^^^
+
+* ``system.flush_metadata_cache()``
+
+  Flush JDBC Metadata caches.
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/jdbc-procedures.fragment:6]: We should add an example usage since people will obviously ask how to use this when multiple catalogs are present.

Something like:

```rst
For example the below query will flush the metadata caches for the `jdbc_catalog_one`.

.. code-block:: sql

    USE jdbc_catalog_one.myschema;
    CALL system.flush_metadata_cache();
```

cc: @mosabua for his inputs.

**Review by hashhar — COMMENTED** [review]
Looks good % comments.

**Comment by hashhar** [comment]
Added the `syntax-needs-review` label (since this PR was broken off from https://github.com/trinodb/trino/pull/10251).

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/IdentifierMappingModule.java (hunk 6 above)
Comment (aczajkowski) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/IdentifierMappingModule.java:61]: Added.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java (hunk 4 above)
Comment (aczajkowski) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:46]: Yes done - leftover from experiments.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java [hunk 8]
```diff
@@ -32,11 +38,52 @@ protected QueryRunner createQueryRunner()
...
+    @Test
+    public void testFlushMetadataCacheProcedureFlushIdentifierMapping()
+    {
+        assertUpdate("CREATE TABLE cached_name AS SELECT * FROM nation", 25);
+
+        String query = "SELECT name, regionkey FROM cached_name";
+        assertQuerySucceeds(query);
+
+        // Rename to lowercase name outside Trino (H2 stores unquoted names as uppercase)
+        h2SqlExecutor.execute("ALTER TABLE tpch.CACHED_NAME RENAME TO tpch.\"cached_name\"");
```
Comment (hashhar) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:76]: Since we used quoted name for the target table we should quote the original name too to make it clearer that the original table had an uppercase name.

```suggestion
        h2SqlExecutor.execute("ALTER TABLE tpch.\"CACHED_NAME\" RENAME TO tpch.\"cached_name\"");
```

Another alternative is to `verify` our assumption that the table has an uppercase name - but doesn't seem worth it (since we need to add a method that returns result set from H2 queries).

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/IdentifierMappingModule.java (hunk 6 above)
Comment (hashhar) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/IdentifierMappingModule.java:63]: Thanks. This however feels a bit odd to do.

@kokosing Do we prefer this or should we add an `invalidate` method to `IdentifierMapping` so that we don't need optional binding to a concrete impl (most impls. will implement a no-op `invalidate` method)?

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java (hunk 8 above)
Comment (aczajkowski) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:76]: I've added verify that actual table stored in H2 is uppercase by checking if query for lowercase fails and uppercase succeed's .

**Comment by aczajkowski** [comment]
> Thanks. This however feels a bit odd to do.

@hashhar @kokosing 
Yes. I thought about NOOP method. But I wanted to keep method names aligned with other caches and keep them `flushCache`. And that looks strange on lvl. of `IdentifierMapping` . 
So i keep it aligned with how it's done with Hive connector where we have similar mechanism. 
All bindings are done to root interface in our case IdentifierMapping so end users don't care if object is cached or not.
And Cached version is optional, providing info on cache enablement or not as it's a rare use case to bind Cached version directly.
Let me know which option you prefere here :) ?
I vote to keep it aligned with Hive.

**Comment by aczajkowski** [comment]
@hashhar how can we proceed with that. Is there anything more should adjust ?
@mosabua - please check against doc changes. If everything is fine from your PoV

File: docs/src/main/sphinx/connector/jdbc-procedures.fragment (hunk 7 above)
Comment (aczajkowski) [reply docs/src/main/sphinx/connector/jdbc-procedures.fragment]: Applied

**Review by hashhar — APPROVED** [review]
I'm fine with the current impl. Seems similar to what we did for Hive.

Looks good to me.

**Comment by hashhar** [comment]
@wendigo @grantatspothero Can you PTAL?

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java [hunk 9]
```diff
@@ -32,11 +38,60 @@ protected QueryRunner createQueryRunner()
...
+    public void testFlushMetadataCacheProcedureFlushMetadata()
+    {
+        assertUpdate("CREATE TABLE cached AS SELECT * FROM nation", 25);
+        assertQuerySucceeds("SELECT name, regionkey FROM cached");
+
+        // Rename column outside Trino
+        h2SqlExecutor.execute("ALTER TABLE tpch.cached ALTER COLUMN regionkey RENAME TO renamed");
+
+        String renamedColumnQuery = "SELECT name, renamed FROM cached";
+        // Should fail as Trino has old metadata cached
```
Comment (grantatspothero) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:58]: The caching happens in two separate places:
1. CachingIdentifierMapping (schema/table mappings cache)
2. CachingJdbcClient (schema/table/column/stats cache)

`testFlushMetadataCacheProcedureFlushMetadata` tests the column cache (from CachingJdbcClient) and `testFlushMetadataCacheProcedureFlushIdentifierMapping` tests the table mapping cache but does not test the schema mapping cache.

Can you add an additional test for schema renames? I think [H2 supports schema renames](https://www.h2database.com/html/commands.html#alter_schema_rename)

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/IdentifierMappingModule.java (hunk 6 above)
Comment (grantatspothero) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/IdentifierMappingModule.java:63]: +1 on @hashhar's suggestion, it is weird to provide both an optional binding and a non-optional binding

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java (hunk 9 above)
Comment (aczajkowski) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:58]: No problem I can add this test.
But I have doubts if it will not be redundant as:
- column and schema cache is inside Jdbc clent. Both are flushed with same/single method call `flushCache`
- both use same caching lib/implementation
- both are fushed with same method `invalidateAll` on cache level in single `flushCache` call

So we would just assure that flushCache method calls all relevant underlining caches.

Let me know your thoughts.

Maybe i will just extend existing test ?
Ofcourse if schema rename will be possible in H2, i will need to check that.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/IdentifierMappingModule.java (hunk 6 above)
Comment (aczajkowski) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/IdentifierMappingModule.java:63]: Its fully allong `OptionalBinder` docs. 
Also this variant is provided as a first option choice.
https://google.github.io/guice/api-docs/4.2/javadoc/com/google/inject/multibindings/OptionalBinder.html

I know it looks weird at first glimps but i found it preatty clear after a while.

We bind 
- optional to some interface A
- then we bind impl. B to interface A

Optional<A> -> A -> B

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java (hunk 9 above)
Comment (aczajkowski) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java:62]: I've extended existing test to cover column, table and schema cache.
https://github.com/trinodb/trino/blob/64ea4ea5f55396b0a6f792dfc5842d0631b19643/plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcCachingConnectorSmokeTest.java#L70:L98

**Comment by aczajkowski** [comment]
Rebased on 366 release to remove conflict

**Comment by grantatspothero** [comment]
Thanks for adding the test, looks good to me 👍

**Comment by aczajkowski** [comment]
> Thanks for adding the schema test, looks good to me 👍 

@grantatspothero Thx for review.

File: docs/src/main/sphinx/connector/jdbc-procedures.fragment [hunk 10]
```diff
@@ -0,0 +1,12 @@
+Procedures
+^^^^^^^^^^
+
+* ``system.flush_metadata_cache()``
+
+  Flush JDBC Metadata caches. For example below query will flush
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/jdbc-procedures.fragment:6]: metadata caches. For example, the following system call flushes the metadata caches for the myschema schema in the example catalog

File: docs/src/main/sphinx/connector/jdbc-procedures.fragment [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+  the metadata caches for the ``jdbc_catalog_one``.
+
+  .. code-block:: sql
+
+      USE jdbc_catalog_one.myschema;
```
Comment (mosabua) [inline docs/src/main/sphinx/connector/jdbc-procedures.fragment:11]: change to 

USE example.myschema

File: docs/src/main/sphinx/connector/jdbc-procedures.fragment (hunk 10 above)
Comment (mosabua) [reply docs/src/main/sphinx/connector/jdbc-procedures.fragment:6]: Also ... is that true or does it flush the caches for all schemas?

**Review by mosabua — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/connector/jdbc-procedures.fragment (hunk 10 above)
Comment (aczajkowski) [reply docs/src/main/sphinx/connector/jdbc-procedures.fragment:6]: Yes procedure is registered in context of single catalog @findepi can you confirm to have 100%

File: docs/src/main/sphinx/connector/jdbc-procedures.fragment (hunk 11 above)
Comment (aczajkowski) [reply docs/src/main/sphinx/connector/jdbc-procedures.fragment]: Applied

File: docs/src/main/sphinx/connector/jdbc-procedures.fragment (hunk 10 above)
Comment (findepi) [reply docs/src/main/sphinx/connector/jdbc-procedures.fragment]: yes, all schemas within a catalog

**Review by findepi — APPROVED** [review]

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/CachingIdentifierMapping.java [hunk 12]
```diff
@@ -71,6 +71,12 @@ public CachingIdentifierMapping(
         this.baseJdbcClient = requireNonNull(baseJdbcClient, "baseJdbcClient is null");
     }
 
+    public void flushCache()
```
Comment (wendigo) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/CachingIdentifierMapping.java:74]: shouldn't it be synchronized?

**Review by wendigo — APPROVED** [review]
LGTM % question:

Is the procedure safe to be exposed to every user?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/CachingIdentifierMapping.java (hunk 12 above)
Comment (findepi) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/mapping/CachingIdentifierMapping.java:74]: it shouldn't need to be

**Comment by aczajkowski** [comment]
> LGTM % question:
> 
> Is the procedure safe to be exposed to every user?

@wendigo  Potentially at some point we might implement access control over that - that good idea. 
But definitely separate PR. It's not operation that might affect physical data, only unwanted side effect that might occur is performance, like there would be no cache.
